family-dashboard/api/src/main/java/org/phenotips/familydashboard/internal/OmimToHpoMapper.java

```diff
@@ -0,0 +1,248 @@
+     *
+     * @param symptom the actual symptom that is associated with an OMIM disorder
+     * @return true if a match between any of the above 3 cases are found, and false otherwise
+     */
+    private boolean hasRelevantPhenotype(String symptom)
+    {
+        if (this.hpoTerms.contains(symptom)) {
+            setCurrentMatch(this.hpoService.getTerm(symptom));
+            return true;
+
+        } else if (foundPatientPhenotypesToSymptomParentsMatch(symptom)) {
+            return true;
+
+        } else if (foundSymptomToPatientPhenotypeParentsMatch(symptom)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Checks for a match between the parent terms of the actual_symptom with any item in the patient's phenotype set.
+     *
+     * @param symptom the actual symptom that is associated with an OMIM disorder
+     * @return
+     */
+    private boolean foundPatientPhenotypesToSymptomParentsMatch(String symptom)
+    {
+        VocabularyTerm symptomTerm = this.hpoService.getTerm(symptom);
+        if (symptomTerm != null) {
+            for (VocabularyTerm term : symptomTerm.getParents()) {
+                System.out.println(term.getId() + ": " + term.getName());
+                if (this.hpoTerms.contains(term.getId())) {
+                    setCurrentMatch(term);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks for a match between the actual_symptom and the parent terms of each item in the patient's phenotype set.
+     *
+     * @param symptom the actual symptom that is associated with an OMIM disorder
+     * @return
+     */
+    private boolean foundSymptomToPatientPhenotypeParentsMatch(String symptom)
+    {
+        for (String feature : this.hpoTerms) {
+            VocabularyTerm patientTerm = this.hpoService.getTerm(feature);
+            if (patientTerm != null) {
+
+                for (VocabularyTerm parent : patientTerm.getParents()) {
+                    System.out.println(parent.getId() + ": " + parent.getName());
+                    if (symptom.equals(parent.getId())) {
+                        setCurrentMatch(patientTerm);
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    private void setOmimToHpoMapping()
+    {
+        this.omimToHpoMapping = new HashMap<>();
+        for (String term : this.omimTerms) {
+            JSONObject omimResult = getOmimTermResults(term);
+            if (omimResult != null) {
+                JSONArray actualSymptoms = omimResult.getJSONArray("actual_symptom");
+                int threshold = actualSymptoms.length() <= this.matchThreshold ?
+                    actualSymptoms.length() : this.matchThreshold;
+                Map<String, JSONObject> matchedFeatures = new HashMap<>();
+                for (int i = 0; i < this.matchThreshold; i++) {
```

Comment:
Hm, actually, if you also do the other change, looking at ancestors with a maximum distance, you should loop over all symptoms, and instead of building a simple map, build a `PriorityQueue`, sorting the closest terms first, then at the end get the first terms from the queue, in order of their priority.